### PR TITLE
[mono] Add support for UnmanagedCallersOnlyAttribute

### PR DESCRIPTION
**Pull request by lambdageek**
When we see a method tagged with `UnmanagedCallersOnlyAttribute` in an `ldtn` instruction, load a pointer to a native-to-managed wrapper instead of a pointer to the managed method.

See https://github.com/dotnet/csharplang/blob/master/proposals/functionpointers.md#systemruntimeinteropservicesunmanagedcallersonlyattribute

TODO:
- [x] Throw exceptions for all the unsupported scenarios:
   - Attempting to call by creating a delegate
   - Invoking via reflection
   - Using unsupported argument or return types
- [x] ~Figure out why the CoreCLR tests are either not running or somehow passing.~
   ~The tests are in https://github.com/dotnet/runtime/tree/master/src/coreclr/tests/src/Interop/UnmanagedCallersOnly which doesn't seem to be in https://github.com/dotnet/runtime/blob/master/src/coreclr/tests/issues.targets but in that case Mono should be failing a number of the tests in there on CI.~
   It's because we only run P0 tests on CI for Mono.  I temporarily set it to P0 for this PR just to watch the tests fail.
- [x] Revert tests back to P1

### CODE REVIEW

File: src/mono/mono/mini/aot-runtime.c [hunk 1]
```diff
@@ -1313,8 +1313,10 @@ decode_method_ref_with_target (MonoAotModule *module, MethodRef *ref, MonoMethod
 			if (!m)
 				return FALSE;
 			klass = decode_klass_ref (module, p, &p, error);
+#if 0
```
Comment (vargaz) [inline src/mono/mono/mini/aot-runtime.c:1316]: Would be better to explicitly encode a NULL class using an extra byte.

**Comment by vargaz** [comment]
Looks ok.

File: src/mono/mono/metadata/marshal.c [hunk 2]
```diff
@@ -6755,3 +6769,34 @@ get_marshal_cb (void)
...
+gboolean
+mono_method_has_unmanaged_callers_only_attribute (MonoMethod *method)
+{
+	/* TODO: if this is slow and called repeatedly for the same method, add a bit to MonoMethod that is true if the
+	 * method doesn't have the UnmanagedCallersOnlyAttribute, and false if we haven't checked yet, or if the
+	 * attribute is present. */
+	ERROR_DECL (attr_error);
+	MonoClass *attr_klass = NULL;
+#ifdef ENABLE_NETCORE
+	attr_klass = = mono_class_try_get_unmanaged_callers_only_attribute_class ();
```
Comment (lambdageek) [inline src/mono/mono/metadata/marshal.c:6787]:
```suggestion
	attr_klass = mono_class_try_get_unmanaged_callers_only_attribute_class ();
```

File: src/mono/mono/mini/jit-icalls.c [hunk 3]
```diff
@@ -39,6 +39,18 @@ mono_ldftn (MonoMethod *method)
 	gpointer addr;
 	ERROR_DECL (error);
 
+	if (G_UNLIKELY ((method->flags & METHOD_ATTRIBUTE_PINVOKE_IMPL) == 0 &&
+			method->wrapper_type == MONO_WRAPPER_NONE &&
+			mono_method_has_unmanaged_callers_only_attribute (method))) {
```
Comment (lambdageek) [inline src/mono/mono/mini/jit-icalls.c:44]: There's a perf problem here
https://github.com/mono/mono/pull/20060#discussion_r449288732

It's actually worse for the JIT than for the interpreter (interp adds the wrapper at transform time, not at execution time; the JIT does this attribute lookup every time `mono_ldftn` gets called at execution time).

I think we can improve things a bit by checking for the attribute in `method-to-ir` and calling a slightly different icall that unconditionally returns the addr of a wrapper.

File: src/mono/mono/metadata/marshal.c [hunk 4]
```diff
@@ -6755,3 +6770,34 @@ get_marshal_cb (void)
...
+
+/**
+ * mono_method_has_unmanaged_callers_only_attribute:
+ *
+ * Returns \c TRUE if \p method has the \c UnmanagedCallersOnlyAttribute
+ */
+gboolean
+mono_method_has_unmanaged_callers_only_attribute (MonoMethod *method)
+{
+	/* TODO: if this is slow and called repeatedly for the same method, add a bit to MonoMethod that is true if the
```
Comment (alexischr) [inline src/mono/mono/metadata/marshal.c:6782]: Just a reminder to remove this comment as you moved the invocation to method_to_ir() so it (probably) doesn't apply anymore

**Review by alexischr — COMMENTED** [review]
LGTM, documenting that this will increase AOT binaries sizes by 1 byte per native-to-managed wrapper AFAICT

**Comment by lambdageek** [comment]
> this will increase AOT binaries sizes by 2 bytes per native-to-managed wrapper AFAICT

1 byte, I think - AOT uses the same value encoding as CIL images - small values (<127) encode to one byte.

**Comment by lambdageek** [comment]
Should pass the positive tests now.  Negative tests will still fail.

**Comment by lambdageek** [comment]
I think preventing calls via `calli` will be difficult using Mono's current design - the native to managed wrapper can't really tell it's being called from a managed method.

File: src/mono/mono/mini/method-to-ir.c [hunk 5]
```diff
@@ -11106,6 +11106,17 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
...
 			}
 
+			/* UnmanagedCallersOnlyAttribute means ldftn should return a method callable from native */
+			if (G_UNLIKELY ((cmethod->flags & METHOD_ATTRIBUTE_PINVOKE_IMPL) == 0 &&
+					cmethod->wrapper_type == MONO_WRAPPER_NONE &&
+					mono_method_has_unmanaged_callers_only_attribute (cmethod))) {
+				MonoClass *delegate_klass = NULL;
+				MonoGCHandle target_handle = 0;
+				cmethod = mono_marshal_get_managed_wrapper (cmethod, delegate_klass, target_handle, cfg->error);
+				CHECK_CFG_ERROR;
```
Comment (lambdageek) [inline src/mono/mono/mini/method-to-ir.c:11116]: This isn't quite right.  On Mono we will throw when JITing the method that has the `ldftn` instruction that has a disallowed `cmethod` (e.g. an instance method).  On CoreCLR they throw when the method runs.  So we would need to capture the error and emit a throw here.

And probably something similar in the interpreter

**Comment by lambdageek** [comment]
Some progress on the negative tests. The InvalidProgramException ones should be in good shape.

**Comment by lambdageek** [comment]
The `TestUnmanagedCallersOnlyValid` test is failing with the interpreter due to https://github.com/dotnet/runtime/issues/38897

**Comment by lambdageek** [comment]
JIT is passing all the tests except `TestPInvokeMarkedWithUnmanagedCallersOnly`.  Working on that one now.

**Comment by lambdageek** [comment]
Once I see the JIT pass all the tests, I'll bump the test back to Pri1 and then the PR is basically complete.

**Comment by lambdageek** [comment]
From https://helix.dot.net/api/2019-06-17/jobs/51ac422f-058f-4fa3-aa34-f41bb363a4fb/workitems/Interop/console the "runtime (Mono Pri0 Runtime Tests Run OSX x64 release)" JIT job:
```
Interop.UnmanagedCallersOnly.XUnitWrapper           Total:  1, Errors: 0, Failed: 0, Skipped: 0, Time: 0.252s
```

As previously mentioned, the interpreter job is failing due to #38897, so this is the best we'll get for now.

File: src/mono/mono/metadata/marshal.c [hunk 6]
```diff
@@ -3940,10 +3960,55 @@ emit_managed_wrapper_noilgen (MonoMethodBuilder *mb, MonoMethodSignature *invoke
 }
 #endif
 
+static gboolean
+type_is_blittable (MonoType *type)
+{
+	if (type->byref)
```
Comment (CoffeeFlux) [inline src/mono/mono/metadata/marshal.c:3966]: Maybe also check for `mono_type_is_reference (type)`?

**Review by CoffeeFlux — APPROVED** [review]
Lots of questions but looks fine to me. Would be really good if Zoltan gets a chance to look at this before it's merged since I'm not terribly confident in my review of the code generation parts. Thanks!

File: src/mono/mono/metadata/marshal.c [hunk 7]
```diff
@@ -3661,6 +3662,25 @@ mono_marshal_get_native_wrapper (MonoMethod *method, gboolean check_exceptions,
...
+		/* emit a wrapper that throws a NotSupportedException */
+		get_marshal_cb ()->mb_emit_exception (mb, "System", "NotSupportedException", "Method canot be marked with both  DllImportAttribute and UnmanagedCallersOnlyAttribute");
+
+		info = mono_wrapper_info_create (mb, WRAPPER_SUBTYPE_NONE);
+		info->d.managed_to_native.method = method;
+
+		csig = mono_metadata_signature_dup_full (get_method_image (method), sig);
+		csig->pinvoke = 0;
+		res = mono_mb_create_and_cache_full (cache, method, mb, csig,
+											 csig->param_count + 16, info, NULL);
```
[the remark this answers is not in the document]
Comment (lambdageek) [reply src/mono/mono/metadata/marshal.c:3677]: I don't know. I'm cargo-culting the other `mono_mb_create_and_cache_full` calls in this function.

File: src/mono/mono/metadata/marshal.c [hunk 8]
```diff
@@ -6755,3 +6843,31 @@ get_marshal_cb (void)
...
+ *
+ * Returns \c TRUE if \p method has the \c UnmanagedCallersOnlyAttribute
+ */
+gboolean
+mono_method_has_unmanaged_callers_only_attribute (MonoMethod *method)
+{
+	ERROR_DECL (attr_error);
+	MonoClass *attr_klass = NULL;
+#ifdef ENABLE_NETCORE
+	attr_klass = mono_class_try_get_unmanaged_callers_only_attribute_class ();
```
[the remark this answers is not in the document]
Comment (lambdageek) [reply src/mono/mono/metadata/marshal.c:6858]: yes, if the attribute is not used anywhere, I think.  So we call the `_try_get_` function and return FALSE if it returns NULL.

[the remark this answers is not in the document]
Comment (CoffeeFlux) [reply src/mono/mono/metadata/marshal.c:6858]: Ah okay, you're using `_try_get_`. Should be good then.

File: src/mono/mono/metadata/marshal.c [hunk 9]
```diff
@@ -3940,10 +3960,55 @@ emit_managed_wrapper_noilgen (MonoMethodBuilder *mb, MonoMethodSignature *invoke
...
+	case MONO_TYPE_U2:
+	case MONO_TYPE_I4:
+	case MONO_TYPE_U4:
+	case MONO_TYPE_I8:
+	case MONO_TYPE_U8:
+	case MONO_TYPE_R4:
+	case MONO_TYPE_R8:
+	case MONO_TYPE_I:
+	case MONO_TYPE_U:
+	case MONO_TYPE_CHAR:
```
Comment (lambdageek) [inline src/mono/mono/metadata/marshal.c:3982]: I think char shouldn't be here. Doesn't look like it's supposed to be blittable. (`mono_class_setup_mono_type` doesn't set the blittable bit on it)

File: src/mono/mono/metadata/marshal.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (CoffeeFlux) [reply src/mono/mono/metadata/marshal.c:3677]: Ugh, that's what I feared. @vargaz any idea? For my own learning if nothing else :)

File: src/mono/mono/metadata/marshal.c (hunk 6 above)
Comment (lambdageek) [reply src/mono/mono/metadata/marshal.c:3966]: I think I want to look at the `MonoClass:blittable` bit for all of those.  This is just to get the `int&` sort of types.

File: src/mono/mono/metadata/marshal.c (hunk 9 above)
Comment (CoffeeFlux) [reply src/mono/mono/metadata/marshal.c:3982]: Ah yeah, good catch. Yes, it's definitely not for the same reason as strong—ambiguity over converting to ANSI, UTF-8, or UTF-16.

File: src/mono/mono/metadata/marshal.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (vargaz) [reply src/mono/mono/metadata/marshal.c:3675]: No idea, it probably comes from the fact that these are wrappers so the deepest stack will be calling the 'wrapped' method, so param_count + <const> is good for max stack.

File: src/mono/mono/metadata/marshal.c [hunk 10]
```diff
@@ -3940,10 +3958,54 @@ emit_managed_wrapper_noilgen (MonoMethodBuilder *mb, MonoMethodSignature *invoke
 }
 #endif
 
+static gboolean
+type_is_blittable (MonoType *type)
```
Comment (vargaz) [inline src/mono/mono/metadata/marshal.c:3962]: There is already a class->blittable flag.

Comment (CoffeeFlux) [reply src/mono/mono/metadata/marshal.c:3962]: We discussed this, but it's not the same because of byref types.

**Pull request merged**